### PR TITLE
Introduce table hierarchical datatype

### PR DESCRIPTION
**Pull request by atmgrifter00**
# Pull Request

## 🤨 Rationale

- #861

This is a staging PR in that it will lay the type groundwork that we can ultimately leverage for representing hierarchical data in the table. There are _no_ functional changes in this PR.

## 👩‍💻 Implementation

Creating a new datatype to represent hierarchical data, which Tanstack requires in order to visualize hierarchical data properly. This type is used for the internal representation of the table data, and thus does not affect our public facing datatype used in the `setData` method. This new datatype basically wraps each record in another object, and also provides another field typed as an array of the new datatype (providing hierarchy).

## 🧪 Testing

Ensured existing tests pass.

## ✅ Checklist

- [ ] I have updated the project documentation to reflect my changes or determined no changes are needed.

### CODE REVIEW

File: packages/nimble-components/src/table/index.ts
Comment (atmgrifter00) [reply packages/nimble-components/src/table/index.ts:215]: Sure. I'm also open to just a more succinct name for this type. Any suggestions?

Comment (mollykreis) [reply packages/nimble-components/src/table/index.ts:215]: I'm struggling to come up with a name that makes sense and isn't too easily confused with our existing types. It's not much more succinct, but maybe something like `TypedInternalTableRecord` would at least help avoid the nested generics.

Comment (atmgrifter00) [reply packages/nimble-components/src/table/index.ts:215]: Turns out we can't easily create a type alias within a class definition, so we're just moving ahead with what's there for now.

**Review by jattasNI — APPROVED** [review]

File: change/@ni-nimble-components-376b2ed8-09ef-409f-9acb-e7f04b283d39.json
Comment (jattasNI) [inline change/@ni-nimble-components-376b2ed8-09ef-409f-9acb-e7f04b283d39.json:1]: Any specs that need updating to capture this concept?

File: packages/nimble-components/src/table/types.ts
Comment (rajsite) [inline packages/nimble-components/src/table/types.ts:52]: how would you feel about calling this `originalRecord`, `userRecord`, `providedRecord`?
I think I like `userRecord` the most. It's a lot clearer to me what the context is when reading a test like the following:
```ts
 const data = [
      { data: { stringField: 'value-1', numberField: 10 } },
      { data: { stringField: 'value-2', numberField: 11 } }
  ];

// vs

 const data = [
    { userRecord: { stringField: 'value-1', numberField: 10 } },
    { userRecord: { stringField: 'value-2', numberField: 11 } }
];
```
Also could we do some typing around saying it is readonly?

File: packages/nimble-components/src/table/types.ts
Comment (rajsite) [inline packages/nimble-components/src/table/types.ts:48]: I'm guessing we can save index on InternalTableRecords in the future right? like `userIndex` for the `userRecord`?

File: packages/nimble-components/src/table/index.ts
Comment (rajsite) [inline packages/nimble-components/src/table/index.ts:271]: shouldn't need the as cast, the fn provided to the map should just specify a return type

File: packages/nimble-components/src/table/types.ts
Comment (jattasNI) [reply packages/nimble-components/src/table/types.ts:52]: I like the idea of a more descriptive name but have a minor preference not to use the term "user" since I'm on a fruitless crusade to use language that differentiates between end-users and clients. I'd vote `originalRecord` or `clientRecord`.

File: packages/nimble-components/src/table/types.ts
Comment (jattasNI) [inline packages/nimble-components/src/table/types.ts:48]: Another possible name for this type would be `TableNode` or `TableRecordNode` to convey that it's a node in a tree. I think I like that slightly better because it conveys its purpose rather than its visibility (internal). But it's your call and I'm fine with the current name.

File: packages/nimble-components/src/table/types.ts
Comment (rajsite) [reply packages/nimble-components/src/table/types.ts:52]: Think in this context I'd lean more to `clientRecord`. Seems more indicative of the thing that came from outside than `originalRecord`

File: packages/nimble-components/src/table/types.ts
Comment (rajsite) [reply packages/nimble-components/src/table/types.ts:48]: Based on the other convo `clientRecord` and `clientRecordIndex` :)

Comment (atmgrifter00) [reply packages/nimble-components/src/table/types.ts:48]: I don't think we can do that through this mechanism unfortunately. This interface needs to mirror what the 3rd party `arrayToTree` method produces. I handled tracking the index in the prototype by keeping a map that associated row Id with a set of metadata, one piece of which was the original index in the array.

Assuming there is more to discuss on this topic, my understanding is that this doesn't need to hold up this PR.

File: change/@ni-nimble-components-376b2ed8-09ef-409f-9acb-e7f04b283d39.json
Comment (atmgrifter00) [reply change/@ni-nimble-components-376b2ed8-09ef-409f-9acb-e7f04b283d39.json:1]: Added some docs to the table data API spec.

File: packages/nimble-components/src/table/types.ts
Comment (atmgrifter00) [reply packages/nimble-components/src/table/types.ts:48]: Went with `TableNode` as that saved me four whole syllables!

File: packages/nimble-components/src/table/types.ts
Comment (jattasNI) [inline packages/nimble-components/src/table/types.ts:44]:
```suggestion
 * Describes a hierarchical data structure that is used for
```

**Review by jattasNI — APPROVED** [review]

**Comment by jattasNI** [comment]
@rajsite @mollykreis I told @atmgrifter00 we can bypass your final review on this one since you're OOO and there will be opportunities for more feedback on the follow up PRs

**Pull request merged**
